lightning/src/chain/mod.rs

```diff
@@ -1,5 +1,100 @@
+use chain::transaction::OutPoint;
+use ln::channelmonitor::{ChannelMonitor, ChannelMonitorUpdate, ChannelMonitorUpdateErr, HTLCUpdate};
+
 pub mod chaininterface;
 pub mod transaction;
 pub mod keysinterface;
+
+/// The `Access` trait defines behavior for accessing chain data and state, such as blocks and UTXO.
+pub trait Access: Send + Sync {
+	/// Returns the transaction output of a funding transaction encoded by [`short_channel_id`].
+	/// Returns an error if `genesis_hash` is for a different chain or if such a transaction output
+	/// is unknown.
+	///
+	/// [`short_channel_id`]: https://github.com/lightningnetwork/lightning-rfc/blob/master/07-routing-gossip.md#definition-of-short_channel_id
+	fn get_utxo(&self, genesis_hash: &BlockHash, short_channel_id: u64) -> Result<TxOut, AccessError>;
+}
+
+/// An error when accessing the chain via [`Access`].
+///
+/// [`Access`]: trait.Access.html
+#[derive(Clone)]
+pub enum AccessError {
+	/// The requested chain is unknown.
+	UnknownChain,
+
+	/// The requested transaction doesn't exist or hasn't confirmed.
+	UnknownTx,
+}
+
+/// The `Watch` trait defines behavior for watching on-chain activity pertaining to channels as
+/// blocks are connected and disconnected.
+///
+/// Each channel is associated with a [`ChannelMonitor`]. Implementations are responsible for
+/// maintaining a set of monitors and updating them accordingly as channel state changes and HTLCs
+/// are resolved on chain. See method documentation for specific requirements.
+///
+/// TODO: Add documentation about persisting monitors.
+///
+/// If an implementation maintains multiple instances of a channel's monitor (e.g., by using a
+/// watchtower), then it must ensure that updates are applied across all instances. Otherwise, it
+/// could result in a revoked transaction being broadcast, allowing the counterparty to claim all
+/// funds in the channel.
+///
+/// [`ChannelMonitor`]: ../ln/channelmonitor/struct.ChannelMonitor.html
+pub trait Watch: Send + Sync {
+	/// Keys needed by monitors for creating and signing transactions.
+	type Keys: ChannelKeys;
+
+	/// Watches a channel identified by `funding_txo` using `monitor`.
+	///
+	/// Implementations are responsible for watching the chain for the funding transaction along
+	/// with spends of its output and any outputs returned by [`get_outputs_to_watch`]. In practice,
+	/// this means calling [`block_connected`] and [`block_disconnected`] on the monitor and
+	/// including all such transactions that meet this criteria.
+	///
+	/// [`get_outputs_to_watch`]: ../ln/channelmonitor/struct.ChannelMonitor.html#method.get_outputs_to_watch
+	/// [`block_connected`]: ../ln/channelmonitor/struct.ChannelMonitor.html#method.block_connected
+	/// [`block_disconnected`]: ../ln/channelmonitor/struct.ChannelMonitor.html#method.block_disconnected
+	fn watch_channel(&self, funding_txo: OutPoint, monitor: ChannelMonitor<Self::Keys>) -> Result<(), ChannelMonitorUpdateErr>;
```

Comment:
Hmm, this just feels like such strange placement - we're returning a unitary enum type from ln in chain. Maybe move ChannelMonitorUpdateErr here or just move all of the on-chain stuff from ln::channelmonitor to chain::channelmonitor?

Reply:
Yeah, I was thinking the same. Leaning towards moving to `chain::channelmonitor`. How are we distinguishing between what goes in `chain` vs `ln`?

Any other PRs to land before making the move? May be worth waiting for a follow-up if so.

Reply:
I dont think there's any ongoing work on this, but historically `ln` was "everything", and there wasn't anything else. Moving all of the channelmonitor stuff to `chain` somehow sounds good to me, and it would clean up this PR a bit, I think, especially to effectively move all of its changes into chain.

Reply:
SGTM. Just to be clear, this is largely just moving the entire `channelmonitor.rs` file from `ln` to `chain`.

Reply:
So, it appears this will be more involved since there are a lot of types, functions, and fields with `pub(super)` visibility either in or used by `channelmonitor.rs`. I could make these `pub` or start moving other files. Moving more files isn't very appealing as I think this will have a cascading effect.

Reply:
Hmm, I mean thats probably an indication of something else - if we want users to be able to implement `chain::Watch` themselves, then anything a default implementation calls probably also needs to be pub.

Reply:
I think the problem is largely limited to `ChannelMonitor` and related types rather than `ChainMonitor`. Specifically, moving it causes visibility issues in `Channel` and `ChannelManager`. The opposite direction is primarily that `ChannelMonitor` uses `OnchainTxHandler`, though I haven't looked into it exhaustively yet.

Reply:
Oh, right, hmm. I mean I don't think having to `s/pub(super)/pub(struct)/` is a reason to not move it, though, no?

Reply:
I'm not arguing to not move it but rather trying to figure out the best approach to resolve visibility without moving more files.
- `pub(<struct>)` is not valid visibility
- `pub(in crate::chain::channelmonitor)` gives "visibilities can only be restricted to ancestor modules"
- `pub(crate)` works
- `pub` works

Leaning towards `pub(crate)` over `pub` so as to not open everything to anyone using the crate. Let me know what you prefer or if I'm missing something.

Reply:
Err, right, i meant pub(crate). I think that's generally fine.